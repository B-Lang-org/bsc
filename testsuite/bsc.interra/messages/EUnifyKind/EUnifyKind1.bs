-----------------------------------------------------------------------
-- Project: Bluespec

-- File: EUnifyKind1.bs

-- Author : Nitin Chand Rahul      <ncr@noida.interrasystems.com>

-- Description: This testcase triggers a Kind error (EUnifyKind)

-- Error Message : bsc EUnifyKind1.bs
-- bsc: Compilation errors:
-- "EUnifyKind1.bs", line 40, column 13, Kind error at:
-- Mystruct

-- Expected kind:
-- *

-- Inferred kind:
-- * -> *
-----------------------------------------------------------------------

package EUnifyKind1 (EUnifyKind1(..), mkEUnifyKind1) where

-- import Int

struct  Mystruct z  =
                      xx ::Int 32
                      yy ::Int 32
interface EUnifyKind1 =
    start  :: Int 32 -> Int 32 -> Action
    result :: Int 32


mkEUnifyKind1 :: Module EUnifyKind1
mkEUnifyKind1 =
      module

        w :: Reg (Int 32)
        w <- mkRegU

        x :: Reg (Int 32)
        x <- mkRegU

        y :: Reg (Int 32)
        y <- mkReg 0

        z :: Mystruct
        z.xx :: Reg (Int 32)

        z.yy :: Reg (Int 32)

        rules
          "Swap":
            when x > y, y /= 0
              ==> action
                             let
                                (x,y) = split z
                      x := y
                      y := x

          "Subtract":
            when x <= y, y /= 0
              ==> action
                      y := y - x

        interface
            start ix iy = action
                              x := ix
                              y := iy
                          when y == 0
            result = x when y == 0


