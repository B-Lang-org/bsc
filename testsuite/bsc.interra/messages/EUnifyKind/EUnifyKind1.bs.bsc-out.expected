checking package dependencies
compiling EUnifyKind1.bs
Warning: "EUnifyKind1.bs", line 57, column 33: (P0102)
  Declaration of `x' shadows previous declaration at "EUnifyKind1.bs", line
  41, column 8
Warning: "EUnifyKind1.bs", line 57, column 33: (P0223)
  Definition of `x' is not used.
Warning: "EUnifyKind1.bs", line 57, column 35: (P0102)
  Declaration of `y' shadows previous declaration at "EUnifyKind1.bs", line
  44, column 8
Warning: "EUnifyKind1.bs", line 57, column 35: (P0223)
  Definition of `y' is not used.
Error: "EUnifyKind1.bs", line 47, column 13: (T0025)
  The type `Mystruct' is applied to too few arguments.
